### DISCUSSION TITLE
Implement Hybrid MP4 for recordings

### DISCUSSION THREAD

**Discussion by mkrn** [HaishinKit/HaishinKit.swift#1560] [Idea]
https://obsproject.com/blog/obs-studio-hybrid-mp4 
This article came out explaining how OBS project implemented recording format compatible with MP4 with one big advantage: 
If recording is interrupted then it's playable (doesn't have problem with moov atom).
This is useful if app crashes, phone overheats, or device runs out of space, etc.
If it's easy enough this should be integrated for recordings (seems like muxing is done differently)

**Comment by shogo4405** [comment]
I have read it. It seems difficult as long as we are using AVAssetWriter. Since the OBS source code cannot be used due to the nature of the GPL, I will need to implement it ourselves, which looks like it will take a lot of time.

There was a similar discussion before. I’d like to try it.
https://github.com/shogo4405/HaishinKit.swift/discussions/1189

**Reply by shogo4405** [reply]
I tried AVAssetWriter.movieFragmentInterval. I was able to confirm that the file remains even when the task is killed. You can test it with this PR: https://github.com/shogo4405/HaishinKit.swift/pull/1561

```swift
// Since it is saved in the document directory, salvage the file if it exists.
let moviesDirectory = URL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0])

do {
  let files = try FileManager.contentsOfDirectory(atPath: moviesDirectory)
  for url in files {
      PHPhotoLibrary.shared().performChanges({() -> Void in
          PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: url)
      }, completionHandler: { _, error -> Void in
          try? FileManager.default.removeItem(at: url)
      })
  }
} catch {
}
```
